macOS/DuckDuckGo/WebExtensions/WebExtensionLoader.swift

```diff
@@ -18,39 +18,91 @@
 
 #if WEB_EXTENSIONS_ENABLED
 
+import CryptoKit
+
 @available(macOS 15.4, *)
 protocol WebExtensionLoading: AnyObject {
-
-    func loadWebExtensions(from paths: [String]) async -> [WKWebExtension]
-
+    @discardableResult
+    func loadWebExtension(path: String, into controller: WKWebExtensionController) async throws -> WKWebExtensionContext
+    func loadWebExtensions(from paths: [String], into controller: WKWebExtensionController) async -> [Result<WKWebExtensionContext, Error>]
+    func unloadExtension(at path: String, from controller: WKWebExtensionController) throws
 }
 
 @available(macOS 15.4, *)
 final class WebExtensionLoader: WebExtensionLoading {
 
-    private func loadWebExtension(path: String) async -> WKWebExtension? {
+    enum WebExtensionLoaderError: Error {
+        case failedToCreateURLFromPath(path: String)
+        case failedToFindContextForPath(path: String)
+    }
+
+    func loadWebExtension(path: String, into controller: WKWebExtensionController) async throws -> WKWebExtensionContext {
```

Comment:
We have to think of the context as a loaded extension, which is why it makes sense for the loader to return this instead of just the extension (which isn't very useful on its own).

The extensions can still be referenced from the contexts, and from the controller if necessary.

Reply:
Sounds good 👍 I like it